lib/diagnostics_channel.js

```diff
@@ -0,0 +1,282 @@
+'use strict';
+
+const {
+  ArrayPrototypeIndexOf,
+  ArrayPrototypePush,
+  ArrayPrototypeShift,
+  ArrayPrototypeSplice,
+  ObjectCreate,
+  ObjectSetPrototypeOf,
+  Map,
+  MapPrototypeDelete,
+  MapPrototypeGet,
+  MapPrototypeSet,
+  Promise,
+  PromiseResolve,
+  Symbol,
+  SymbolAsyncIterator
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+const { triggerUncaughtException } = internalBinding('errors');
+
+const { WeakReference } = internalBinding('util');
+
+function identity(v) {
+  return v;
+}
+
+class AsyncIterableChannel {
+  constructor(channel) {
+    this.channel = channel;
+    this.events = [];
+    this.waiting = [];
+
+    this.subscriber = (message) => {
+      const resolve = ArrayPrototypeShift(this.waiting);
+      if (resolve) {
+        return resolve({
+          value: message,
+          done: false
+        });
+      }
+
+      ArrayPrototypePush(this.events, message);
```

Comment:
is there any reason someone might replace the .events property on this?

Reply:
I don't think so, no. Undecided if I should even keep the async iterable feature. Just added as an idea when someone else mentioned it, but not sure how the utility is really worth the extra code, unless maybe channels were made into a more generic construct that got use beyond just for diagnostics channels. 🤔